Documentation for decreases clauses

SaswatPadhi opened the pull request:
Related to #6149.

Thanks to @LongPham7 for an initial draft!

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a ~~Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.~~
- n/a ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- n/a ~~Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).~~
- n/a ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- n/a ~~White-space or formatting changes outside the feature-related changed lines are in commits of their own.~~

SaswatPadhi committed:
Documentation for decreases clauses

Co-authored-by: Long Pham <lopha@amazon.com>

jimgrundy commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,183 @@
 
-TODO: Document `__CPROVER_decreases`
+A _decreases_ clause specifies a measure that must strictly decrease
+monotonically at every iteration of a loop.
```
I would simply say "must strictly decrease at every iteration" rather than "strictly decrease monotonically at every iteration". Monotonically decreasing means decreases or remains the same, but never increasing. Strictly decreasing is stronger so although adding "monotonically" doesn't alter the meaning it's presence only serves to add a bit of doubt as to what you meant.

SaswatPadhi replied:
Yes, "monotonically decreases" / "decreases monotonically" means what you mentioned, but "strictly decreases monotonically" means a measure decreases consistently (doesn't remain the same / increases) and [is a widely used](https://www.google.com/search?q=%22strictly+decreases+monotonically%22) phrase. I don't think it's incorrect, but I agree that it may be confusing, so I'll change this as you suggested.

jimgrundy commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,183 @@
+Unlike mathematical integers, arithmetic types in C are all bounded from below.
+Furthermore, unlike mathematical real numbers, floating-point arithmetic types in C are not dense.
+Hence, any strictly decreasing arithmetic expression in C can be used as a decreases clause.
+
```
Echoing Daniel's comments to Felipe on earlier documents, I wouldn't try to explain the science here. Maybe we write a background document for folks, but people come to a man-page type document like this for a short and sweet description of what a feature does and how should they use it. An explanation of how/why it works is done better in another format.

SaswatPadhi replied:
Thanks; makes sense. This paragraph could be removed.

jimgrundy commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,183 @@
+
+**Important.**
+Like invariant clauses, decreases clauses must be free of side effects,
+for example, mutation of local or global variables.
```
Do we check that? Are we telling people to take care here lest they produce an unsound result, or are we telling them that if they make this mistake the tool will complain with an error. If so, it might be nice to describe the error they will get.

SaswatPadhi replied:
Yes, we do check this and generate an error as shown in this regression test:
https://github.com/diffblue/cbmc/blob/3e0007f9ed818969185676eb96f8811dc75e8705/regression/contracts/variant_side_effects_fail/test.desc#L4

I would add this message to this doc & the invariants doc.

jimgrundy commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,183 @@
+we extend the strict arithmetic comparison to a strict lexicographic comparison.
+Formally, given two evaluations `(a_1, ..., a_n)` and `(b_1, ..., b_n)`,
+a strict lexicographic comparison, denoted by `(a_1, ..., a_n) ≺ (b_1, ..., b_n)`,
+is defined as the predicate:
```
Kind of repeating myself, but I think I would be better served by a separate document that described the unbounded verification feature and how it works (both mathematically and how implemented in CBMC) and keep this document as a light weight description of what it does and how to use it.

SaswatPadhi replied:
I agree on the previous comment, but this one seems relevant to end users. It's not describing the science, or "why" comparison of tuples is defined this way -- there are many different ways one may define a comparison over tuples), but is only describing "how" tuples are compared (i.e., lexicographically) which I think end users must know in order to be able to write multi-dimensional decreases clauses.

feliperodri commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,179 @@
+A _decreases_ clause specifies a measure that must strictly decrease at every iteration of a loop.
+By demonstrating that the measure
+1. is bounded from below, and
+2. strictly decreases at each iteration,
+we can prove termination of loops.
+This is because the measure must eventually hit the lower bound
```
You're appending the last paragraph in `2.`. Add an empty line between them (or write them as a single paragraph).

SaswatPadhi replied:
Good catch! Thanks

feliperodri commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,179 @@
+# Decreases Clauses
 
-TODO: Document `__CPROVER_decreases`
+A _decreases_ clause specifies a measure that must strictly decrease at every iteration of a loop.
```
Do you have a link to any external resources with more details in the theory behind decreases clauses? It could be useful to add it here.

SaswatPadhi replied:
Good point; I'll add some resources.

feliperodri commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,179 @@
+
+### Syntax
+
+A one-dimensional (1D) decreases clause for a loop is an arithmetic expression `e`
+over the variables visible at the same scope as the loop,
+specified as `__CPROVER_decreases(e)`.
```
It doesn't have to be only arithmetic expressions, right? What about `__CPROVER_decreases(i)`?

SaswatPadhi replied:
Numeric variables by themselves are valid arithmetic expressions too -- [any expression whose return type is numeric is an arithmetic expression](https://www.cis.upenn.edu/~matuszek/General/JavaSyntax/arithmetic-expressions.html). `5`, `x` (for numeric `x`), `x+5`, `x <= 0 ? -x : x` are all valid arithmetic expressions.

feliperodri commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,179 @@
+
+### Semantics
+
+A decreases clause extends the loop abstraction introduced in the [invariants clause](contracts-invariants.md) document.
```
```
s/document/manual/g
```

feliperodri commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,179 @@
+**Important.** 
+Decreases clauses work in conjunction with [loop invariants](contract-invariants.md),
+which model an arbitrary loop iteration at which the decreases clause is checked.
+If a decreases clause is annotated on a loop without an invariant clause,
+then the weakest possible invariant (i.e, `true`) is used to model an arbitrary iteration.
```
In this case, a successful verification will always means the loops terminates, but can't guarantee memory safety of the implementation?

SaswatPadhi replied:
Yes, exactly

feliperodri commented:
@SaswatPadhi can we close #6149 after merging this PR? The only initial documentation missing will be assigns clause for loops, correct?

SaswatPadhi committed:
binary_search regression test for loop contracts

SaswatPadhi commented:
> @SaswatPadhi can we close #6149 after merging this PR? The only initial documentation missing will be assigns clause for loops, correct?

I think we can close it after this + assigns clause documentation is merged. I would push the assigns clause documentation to #6249

SaswatPadhi commented:
I have addressed all prior comments and fixed some build issues (with the new binary_search regression test).

Keeping this PR open for one final review from @jimgrundy.

jimgrundy commented on doc/cprover-manual/contracts-decreases.md:
````diff
@@ -1,3 +1,184 @@
+|| (a_1 == b_1 && a_2 < b_2)
+|| ...
+|| (a_1 == b_1 && ... && a_n < b_n)
+```
````
I would suggest cutting the material after "strict lexicographic comparison". Maybe hyperlink off "lexicographic comparison" to the article you linked earlier, and you can skip the explanation of how you implement a lexicographic compare. If people know what one is already they don't need to see this code. If they don't, they might be better off reading the code. But, if you like it, then I'm OK with it.

SaswatPadhi replied:
Okay, I could remove it. I wasn't sure if end users would be familiar with exactly what lexicographic comparison on ordered tuples would mean. I didn't want to point them to research papers just for that, but I guess they could Google it if they are unsure :)

jimgrundy commented on doc/cprover-manual/contracts-decreases.md:
```diff
@@ -1,3 +1,184 @@
+  if(size <= 0 || buf == NULL) return NOT_FOUND;
+
+  long long lb = 0, ub = size - 1;
+  long long mid = (lb + ub) / 2;
```
This works because we use "long long" values to avoid overflow. But, we can do the computation another way and avoid overflow while using fewer bits, which might give better solver performance. Take a look at this article for a nice clear way to express the algorithm without the extra bits: [Extra, Extra - Read All About It: Nearly All Binary Searches and Mergesorts are Broken](https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html). Now, since the use of int for the arguments and long long for the computation should mean this is correct (are there machines for which int is the same size as long long?) then I'm fine with leaving this as is, but I think you might like the version in this article better anyway

SaswatPadhi replied:
Thanks! :D Great point!

> Now, since the use of int for the arguments and long long for the computation should mean this is correct (are there machines for which int is the same size as long long?)

It's correct (verified) only on 64-bit Windows and Linux machines -- the CBMC CI machines & my machine.
I originally had `long`, and that failed because on Windows `int` == `long`, apparently! Then I changed to `long long` and both Windows + Linux were happy. But that's still fragile -- even `long long` could be the same as `int` on some platform.

I implemented the change. The verification time with MiniSAT remains the same (~2.7s), but the number of variables in the SAT formula goes from ~12900 down to ~7700.
